lib/fizzy/parser_expr.cpp

```diff
@@ -173,41 +175,74 @@ void push_branch_immediates(
     push(immediates, arity);
 }
 
-inline void mark_frame_unreachable(ControlFrame& frame) noexcept
+inline void mark_frame_unreachable(ControlFrame& frame, Stack<ValType>& operand_stack) noexcept
 {
     frame.unreachable = true;
-    frame.stack_height = frame.parent_stack_height;
+    operand_stack.shrink(static_cast<size_t>(frame.parent_stack_height));
 }
 
+inline void drop_operand(
+    const ControlFrame& frame, Stack<ValType>& operand_stack, std::optional<ValType> expected_type)
+{
+    if (!frame.unreachable &&
+        static_cast<int>(operand_stack.size()) < frame.parent_stack_height + 1)
+        throw validation_error{"stack underflow"};
+
+    if (frame.unreachable && static_cast<int>(operand_stack.size()) == frame.parent_stack_height)
+        return;
+
+    const auto actual_type = operand_stack.pop();
+    if (expected_type.has_value() && actual_type != *expected_type)
+        throw validation_error{"type mismatch"};
+}
+
+inline void push_operand(Stack<ValType>& operand_stack, ValType type)
+{
+    operand_stack.push(type);
+}
+
+ValType find_local_type(
+    const std::vector<ValType>& params, const std::vector<Locals>& locals, LocalIdx idx)
+{
+    if (idx < params.size())
+        return params[idx];
+
+    const auto local_idx = idx - params.size();
+    uint64_t local_count = 0;
+    for (const auto& l : locals)
```

Comment:
This algorithm will have to be replaced. There exist log2 algorithm without needing to expand to vector of individual types.

Reply:
I guess you mean something like iterating over it once to create an array of cumulative sums, then using binary search on this array.

I can add a TODO.

Reply:
Yes, this is what I mean. And I'm also for doing it later.